_posts/2020-06-28-openfaas-loki-into.md

```diff
@@ -0,0 +1,235 @@
+
+### Recreate the port-forward
+Earlier, while we were installing and verifying the system, we ran this command
+
+```bash
+kubectl port-forward -n openfaas svc/gateway 8080:8080 &
+```
+
+Unfortunately, that connection is no longer valid, because we updated the OpenFaaS installation to configure the logs provider.  We need to stop the current `port-forward` command and then start a new one:
+
+```bash
+pkill kubectl
+kubectl port-forward -n openfaas svc/gateway 8080:8080 &
+```
+
+### Invoke the function and check the logs
+Now we can use the `faas-cli` to invoke and see the logs, this time powered by Loki.
+
+```bash
+faas-cli invoke cows <<< ""
+             ___   ___
+             \  \  \  \
+   (___)   ___\__\__\__\__
+   (o o)   |  O O O O O O|
+  --\ /----+-------------+-------/
+  |  O                          /
+   \                           /
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~s~~~~~~~~~~~
+         Steerage
+```
+and
+
+```bash
+faas-cli logs cows
+2020-06-27T18:24:09Z 2020-06-27T18:24:09.633228Z stderr F 2020/06/27 18:24:09 Forking fprocess.
+2020-06-27T18:24:09Z 2020-06-27T18:24:09.8101492Z stderr F 2020/06/27 18:24:09 Wrote 279 Bytes - Duration: 0.176972 seconds
+```
+
+
+To compare with the Kubernetes powered logs
+
+```bash
+2020-06-27T16:37:17Z 2020/06/27 16:37:17 Forking fprocess.
+2020-06-27T16:37:17Z 2020/06/27 16:37:17 Wrote 119 Bytes - Duration: 0.130287 seconds
+```
+
+We see a slightly different format that is coming from the way Loki scrapes the logs. Primarily
+
+```
+2020-06-27T18:24:09.633228Z stderr F
+```
+
+This shows the timestamp for when the log was collected and that it came from `stderr` instead of `stdout`.
+
+We can also demonstrate that the logs are now independently stored from the function instance by scaling the function to 0 and still retrieving the same logs
+
+```bash
+kubectl -n openfaas-fn scale deployment --replicas 0 cows
+faas-cli logs cows
+2020-06-27T18:24:09Z 2020-06-27T18:24:09.633228Z stderr F 2020/06/27 18:24:09 Forking fprocess.
+2020-06-27T18:24:09Z 2020-06-27T18:24:09.8101492Z stderr F 2020/06/27 18:24:09 Wrote 279 Bytes - Duration: 0.176972 seconds
+```
+
+>Note: The `faas-cli logs` defaults to the last 5 mins of logs. If you walk through this tutorial and have a long pause between steps, it is possible that your logs will be empty because of the gap between your invocation and the request.  Just run the invoke command again and then get the logs.
+
+
+## Wrapping up
+
+In this post we have shown how you can use Loki and OpenFaaS together so that your function logs are stored independently from the instances. This should ease your debugging experience with functions and the combination with Grafana also opens some additional possibilities, for example you can start creating dashboards for your functions that combine Prometheus metrics _and_ logs on the same screen by using the [new logs panel](https://grafana.com/docs/grafana/latest/panels/visualizations/logs-panel/)
```

Comment:
Nice summary 👍

Reply:
This sentence is perhaps a little long and could benefit from a `.`

Might also work as `The combination of Loki and Grafana make it even easier to debug functions than before. You can even create dashboards that combine function metrics and logs in one place.`